lib/vrl/stdlib/src/encode_logfmt.rs

```diff
@@ -0,0 +1,227 @@
+        match self.value {
+            Value::Bytes(b) => {
+                let val = String::from_utf8_lossy(b);
+                format_logfmt_string(f, &val)
+            }
+            _ => {
+                let val = format!("{}", self.value);
+                format_logfmt_string(f, &val)
+            }
+        }
+    }
+}
+
+struct LogFmtFieldFormatter<'a> {
+    key: &'a str,
+    value: &'a Value,
+}
+
+impl std::fmt::Display for LogFmtFieldFormatter<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        format_logfmt_string(f, self.key)?;
+        let value_format = LogFmtFieldValueFormatter { value: self.value };
+        write!(f, "={}", value_format)
+    }
+}
+
+struct LogFmtFormatter {
+    value: Value,
+}
+
+impl std::fmt::Display for LogFmtFormatter {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.value {
+            Value::Object(map) => {
+                for (idx, (key, value)) in map.iter().enumerate() {
+                    let field_formatter = LogFmtFieldFormatter { key, value };
+                    if idx > 0 {
+                        f.write_str(" ")?;
+                    }
+                    write!(f, "{}", field_formatter)?;
+                }
+
+                Ok(())
+            }
+            Value::Array(arr) => {
+                for (idx, value) in arr.iter().enumerate() {
+                    if idx > 0 {
+                        f.write_str(" ")?;
+                    }
+
+                    match value {
+                        Value::Array(arr) if arr.len() == 2 => {
+                            let (key, value) = (&arr[0], &arr[1]);
+                            if let Value::Bytes(b) = key {
+                                let key_str = String::from_utf8_lossy(b);
+                                let field_formatter = LogFmtFieldFormatter {
+                                    key: &key_str,
+                                    value,
+                                };
+                                write!(f, "{}", field_formatter)?;
+                            }
+                        }
+                        _ => return Err(fmt::Error),
+                    }
+                }
+
+                Ok(())
+            }
+            _ => Err(fmt::Error),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct EncodeLogfmtFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for EncodeLogfmtFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let formatter = LogFmtFormatter { value };
+
+        let mut output = String::new();
+        match std::fmt::write(&mut output, format_args!("{}", formatter)) {
+            Ok(_) => Ok(output.into()),
+            Err(_) => Err("Failed to encode logfmt".into()),
+        }
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value.type_def(state).infallible().bytes()
```

Comment:
```suggestion
        self.value.type_def(state).fallible().bytes()
```

I think this function will actually always be fallible due to the way the encoding works for arrays where it can fail based on the input.